govultr.go

```diff
@@ -173,14 +179,34 @@ func (c *Client) NewRequest(ctx context.Context, method, uri string, body url.Va
 	return req, nil
 }
 
-// DoWithContext sends an API Request and returns back the response. The API response is checked  to see if it was
-// a successful call. A successful call is then checked to see if we need to unmarshal since some resources
-// have their own implements of unmarshal.
+// DoWithContext handles the retry mechanism while handing off the API request and returning back errors.
 func (c *Client) DoWithContext(ctx context.Context, r *http.Request, data interface{}) error {
 
-	// Sleep this call
+	// Initial sleep on the call
 	time.Sleep(c.RateLimit)
 
+	var err error
+	for tryCount := 1; tryCount <= c.RetryLimit; tryCount++ {
+
+		err = c.executeHTTP(ctx, r, data)
```

Comment:
This is unsafe, as requests with bodies should not be reused. This is because the body reader may get consumed on a rate-limited request, which makes the next tries empty. The request body needs to be duplicated into a buffer which is reset each time.

Reply:
In addition, the context should be wrapped to allow for proper cancellation.

Reply:
If you're fine with adding a dependency, it may be a better idea to use https://github.com/hashicorp/go-retryablehttp. If not, then I can either do the minimum amount of changes to make this PR correct, or I can do a larger refactor of the request code to make it cleaner to implement (but that wouldn't be until later this week at the earliest).

Reply:
By all means we would really appreciate the refactor! I have no issue with adding in that hashicorp dependency if it will improve the client overall. I'll keep this PR open for now and close it out once you are set with your refactored changes.

Thanks!

Reply:
OK, sounds good. I'll start with using the lib for implementing retry as one PR so we cam get that fixed. Afterwards, when I have time, I'll do some refactoring of the request code to make it easier to implement this kind of thing in the future (splitting net/http requests from Vultr requests so stuff can be modified more easily).

Reply:
Awesome! Thanks for the contributions!